toffy/fov_watcher.py

```diff
@@ -0,0 +1,239 @@
+import os
+import time
+import json
+from datetime import datetime
+from typing import Callable, List, Tuple
+from watchdog.events import FileCreatedEvent, FileSystemEventHandler
+from watchdog.observers import Observer
+
+
+class RunStructure:
+    """Expected bin and json files
+
+    Attributes:
+        completion (dict): Whether or not an expected file has been created
+    """
+    def __init__(self, run_folder: str, timeout: int = 10 * 60):
+        """ initializes RunStructure by parsing run json within provided run folder
+
+        Args:
+            run_folder (str):
+                path to run folder
+            timeout (int):
+                number of seconds to wait for non-null filesize before raising an error
+        """
+        self.timeout = timeout
+        self.completion = {}
+
+        # find run .json and get parameters
+        with open(f'{run_folder}.json', 'r') as f:
+            run_metadata = json.load(f)
+
+        # parse run_metadata and populate expected structure
+        for fov in run_metadata.get('fovs', ()):
+            run_order = fov.get('runOrder', -1)
+            scan = fov.get('scanCount', -1)
+            if run_order * scan < 0:
+                raise KeyError(f"Could not locate keys in {run_folder}.json")
+
+            fov_name = f'fov-{run_order}-scan-{scan}'
```

Comment:
essentially yes. I wasn't sure if there'd be a situation where an fov is skipped within the json or scan ever isn't one, so I decided to fill the 'completion' dictionary only with what's explicitly detailed in the json.  i'm open to fixing the ending to 'scan-1' if that's the only relevant case.

Reply:
to clarify, this is the run json, so it would need every file name.  it would still make sense for it to store that but I guess someone at ionpath really loves re-implementing manual filepath construction everytime they interact with the json 🤷